Implements address space cleanup when a workspace service is uninstalled

JC-wk opened the pull request:
# Resolves #4727

## PR

- [x] unit testing may be needed?
- [x] I have noted an edge case of the address being released during a failed uninstall and then allocated elsewhere. I need to move the deallocation process to run after the main step has succeeded so it runs after the terraform uninstall. I am testing doing this in the service_bus now and it is working ok.

## What is being addressed
Currently address spaces are not cleaned up when a workspace service is uninstalled from a workspace, this could lead to ip range exhaustion.
This PR adds the functionality to delete the `address_space` used by a workspace-service on uninstall of the service. The address range is then freed from the workspace and can be reused.

## How is this addressed

- Adds a step to the api to delete the address space when a workspace service is uninstalled
- Updates templates to conduct a workspace upgrade after uninstall
- I have not added any tests but happy to accept advice on how to implement these
- Updated documentation
- Updated CHANGELOG.md
- Increment template versions

JC-wk committed:
Implement address space removal

JC-wk committed:
update template versions

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

JC-wk committed:
Move address_space deletion to service_bus

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

Commit:
Merge branch 'main' into deallocate-ip-addresses

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

Commit:
bump api version

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

Commit:
changelog

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

Commit:
Copilot feedback

JC-wk commented on CHANGELOG.md:
```diff
@@ -4,7 +4,7 @@
 * _No changes yet_
 
 ENHANCEMENTS:
-* _No changes yet_
+* `address_spaces` will now be removed from a workspace when a workspace service that uses an `address_space` is deleted to prevent IP address range exhaustion ([#4727](https://github.com/microsoft/AzureTRE/issues/4727))
 
```
done

JC-wk commented on api_app/service_bus/deployment_status_updater.py:
```diff
@@ -187,6 +190,34 @@ async def update_status_in_database(self, message: DeploymentStatusUpdateMessage
+                                    workspace_patch.properties = {"address_spaces": new_address_spaces}
+                                    try:
+                                        await workspace_repo.patch_workspace(workspace, workspace_patch, workspace.etag, self.resource_template_repo, self.resource_history_repo, operation.user, False)
+                                        logger.info(f"Freed address space {address_to_free} from workspace {parent_workspace_id} after successful uninstall of {resource_id}")
```
done

Commit:
remove spaces

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

JC-wk committed:
Bump version to 0.25.27

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

marrobi commented:
## Review: Address space cleanup on workspace-service uninstall

### Overall
The approach is sound — hooking cleanup to the successful `main` step of an uninstall (so Terraform destroy completes before the address is freed) is the right ordering, and the retry-on-etag-conflict logic is a nice touch. A few correctness and design points below.

### Correctness / logic

**1. Interaction between direct patch and the new `upgrade` step.**
The cleanup block sits *after* the `if is_last_step is False:` branch that enqueues the next step. Since the templates now add a trailing `upgrade workspace` step, when `main` succeeds we both (a) enqueue the upgrade step and (b) directly `patch_workspace` to remove the entry from `address_spaces`. That's two things mutating the workspace's `address_spaces`. Can you clarify the intended design? If the `upgrade` step already reconciles this, the direct patch may be redundant (and vice versa). If both are required, the ordering/interaction should be documented.

**2. Double-nested broad `try/except` swallows failures.**
The cleanup is wrapped in two catch-all `except Exception` blocks that only log. A failed free is therefore invisible — the service reports `Deleted` successfully but the range is silently leaked, which is exactly the bug this PR targets. Suggest collapsing the redundant outer `try/except` and emitting an alertable signal (or a resource note) when cleanup ultimately fails.

**3. `patch_workspace(..., False)` behaviour.**
Please confirm this patch doesn't trigger another workspace deployment/operation. Combined with the added upgrade step, concurrent workspace operations could cause the very etag churn the retry loop is compensating for.

**4. Extract cleanup into a private method.**
The ~35-line block is nested up to 7 levels deep inside the already-long `update_status_in_database`. Pulling it into `_free_workspace_address_space(self, resource_to_persist, operation)` would improve readability and make it directly unit-testable. It'd also let `max_retries = 3` become a module-level constant instead of a magic number.

### Tests
Good coverage (happy path, missing-property parametrization, retry, max-retries). Two suggestions:
- Add a test covering the interaction/ordering between the direct patch and the enqueued `upgrade` step.
- `test_..._logs_error_after_max_retries` patches `services.logging.logger.exception` globally and asserts `assert_called_once_with`, which is brittle. Prefer asserting on the module's specific logger.

### Versioning / changelog
- The description says this affects *any* workspace service using an `address_space`, but only `azureml` and `databricks` got the upgrade step + version bump. Please confirm these are the only bundled services that allocate an `address_space`; otherwise others need the same treatment for full coverage.

JC-wk committed:
Merge branch 'main' into deallocate-ip-addresses

Commit:
### Summary of Changes

  1. Ordering & Interaction between direct patch and enqueued upgrade step (deployment_status_updater.py):
      • Moved _free_workspace_address_space to execute before if is_last_step is False: (which enqueues the next step).
      • patch_workspace directly updates the Workspace resource document in Cosmos DB to remove address_to_free from workspace.
      properties["address_spaces"].
      • By running _free_workspace_address_space first, when update_resource_for_step runs for the trailing workspace upgrade step, it
      queries the workspace document from Cosmos DB with the address space ALREADY removed.
      • The trailing workspace upgrade step then executes Terraform against Azure infrastructure using the updated Cosmos DB state.
  2. Cleaned up try/except error logging (deployment_status_updater.py):
      • Extracted cleanup into _free_workspace_address_space and collapsed double try/except loops.
      • Emits structured, alertable logs logger.error("[ADDRESS_SPACE_CLEANUP_FAILED] ...", exc_info=True) when cleanup exhausts retries
      or fails.
  3. Confirmed patch_workspace(..., False) behavior:
      • patch_workspace with force_version_update=False only updates the database document and records history; it does not trigger a
      new deployment operation or send messages to Service Bus. Added clarifying comments in the code.
  4. Extracted private method and constant (deployment_status_updater.py):
      • Created MAX_CLEANUP_RETRIES = 3 at module scope.
      • Extracted async def _free_workspace_address_space(self, resource_to_persist: dict, operation: Operation).
  5. Updated & added unit tests (test_deployment_status_update.py):
      • Refactored test_workspace_service_uninstall_logs_error_after_max_retries to patch @patch('service_bus.deployment_status_updater.
      logger') and assert on the module logger.
      • Added test_workspace_service_uninstall_frees_address_space_before_enqueuing_next_step to verify patch_workspace runs before
      send_deployment_message.
  6. Full coverage of bundled workspace services (template_schema.json, porter.yaml):
      • Audited all workspace services (azureml, databricks, gitea, guacamole, health-services, mysql, ohdsi, openai, azuresql).
      • Identified that ohdsi also defines an address_space parameter (for PostgreSQL subnet).
      • Added trailing workspace upgrade step to ohdsi's uninstall pipeline and bumped ohdsi porter bundle version to 0.3.8.

  ### Verification

  • Ran full test suite (pytest api_app/tests_ma): 730 passed in 17.86s.

Commit:
copilot feedback

Commit:
• Refactored AutoLockRenewer in deployment_status_updater.py to use an asynchronous context manager (async with AutoLockRenewer() as
  renewer:).
  • This ensures that renewer.close() is guaranteed to execute upon exiting the block, even if an exception occurs while iterating over
  messages or completing/abandoning a message.

Commit:
1. **deployment_status_updater.py**:
      • Captured the active tracing span context via with tracer.start_as_current_span("process_message") as current_span:.
      • Set the attributes step_id, operation_id, and status on current_span after parsing the DeploymentStatusUpdateMessage.
  2. **test_deployment_status_update.py**:
      • Added a unit test test_process_message_sets_span_attributes to verify that set_attribute is called with the expected values.

Commit:
Add SERVICE_BUS_MAX_WAIT_TIME configuration and improve error handling in DeploymentStatusUpdater

Commit:
Fix CosmosAccessConditionFailedError instantiation in tests for airlock request and resource patching

Commit:
Fix type casting for span attributes in DeploymentStatusUpdater and update test assertions
